source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,159 @@
+
+    // |vec_id1|, |vec_id2| and |vec_id3| must be fresh.
+    if(!fuzzerutil::IsFreshId(ir_context, message_.vec_id1()) || !fuzzerutil::IsFreshId(ir_context, message_.vec_id2())
+        || !fuzzerutil::IsFreshId(ir_context, message_.vec_id3())) return false;
+
+    // |vec_id1|, |vec_id2| and |vec_id3| should have disparate ids.
+    if(message_.vec_id1() == message_.vec_id2() || message_.vec_id2() == message_.vec_id3() || message_.vec_id1() == message_.vec_id3()) return false;
+
+    // |pos| needs to be a non-negative integer less than the vector length.
+    auto vec_len = vector_type->GetSingleWordOperand(1);
+    if(message_.scalar_position() >= vec_len) return false;
+
+    auto vec1 = message_.vec1_elements();
+    auto vec2 = message_.vec2_elements();
+    // The vectors being populated must have the same length as specified by vector type.
+    if(vec1.size() != (int)vec_len || vec2.size() != (int)vec_len) return false;
+
+    auto type_id = instruction->type_id();
+    auto matchType = [&ir_context, &type_id](uint32_t id) {return ir_context->get_def_use_mgr()->GetDef(id)->type_id() == type_id;};
+
+    // All ids should match the type_id specified in the instruction.
+    if(!std::all_of(vec1.begin(), vec1.end(), matchType) || !std::all_of(vec2.begin(), vec2.end(), matchType)) return false;
+
+    // Get the OpConstant instruction with id corresponding to index |pos|.
+    auto constant1 = ir_context->get_def_use_mgr()->GetDef(vec1[message_.scalar_position()]);
+    auto constant2 = ir_context->get_def_use_mgr()->GetDef(vec2[message_.scalar_position()]);
+    // The constants at position |pos| of the vectors should be zero constants.
+    if(constant1->GetSingleWordOperand(1) || constant2->GetSingleWordOperand(1)) return false;
+
+    return true;
+}
+
+void TransformationWrapVectorSynonym::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+    // OpCompositeConstructs are inserted before the original arithmetic type instruction.
+    auto instruction = ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+    auto vec1 = message_.vec1_elements();
+    auto vec2 = message_.vec2_elements();
+
+    // Change the target position with variable from the original instruction.
+    vec1[message_.scalar_position()] = instruction->GetSingleWordOperand(1);
+    vec2[message_.scalar_position()] = instruction->GetSingleWordOperand(2);
+
+    auto inst_descriptor = MakeInstructionDescriptor(message_.instruction_id(), instruction->opcode(), 0);
+
+    // Apply transformation to add two composite construct.
+    ApplyTransformation(TransformationCompositeConstruct(
+        message_.vec_type_id(), message_.vec1_elements(), inst_descriptor, message_.vec_id1()));
+    ApplyTransformation(TransformationCompositeConstruct(
+        message_.vec_type_id(), message_.vec1_elements(), inst_descriptor, message_.vec_id2()));
```

Comment:
By design, transformations shouldn't nest (i.e. one transformation shouldn't apply other transformations). This simplifies the transformations and decouples their logic. Thus, if your transformation depends on others being applied, you can apply them in the fuzzer pass. 

Also, the `ApplyTransformation` method is a public method of a `FuzzerPass` class, thus, you can't call it in the transformation class.

Reply:
So what should happen in this transformation though. If I want to test it in the test file it seems like this transformation has to  in some way add two new ``OpCompositeConstruct``.

Reply:
You can add those `OpCompositeConstruct` in the fuzzer pass by using `ApplyTransformation(TransformationCompositeConstruct(...))` and then pass the result ids of the created instructions into this transformation.

Reply:
Done.